Element / Block API: Add RawHTML component, drop support for HTML string block save

aduth opened the pull request:
Related: #3745 (cherry-picks 57978ebebf2d7f7d57a29ccb5bc958a3a946aa42)

This pull request seeks to migrate away from string return from block `save` implementation toward a new `RawHTML` component, which serves an identical purpose while also providing additional utility for nesting (#3745) and Editable value (#2750). Further, it reduces the number of code paths to maintain and improves consistency of hook behaviors (e.g. hooks injecting props via `blocks.getSaveContent.extraProps` are now respected). You can see this demonstrated here in the change introducing `supports.className = false` to the Classic block, as otherwise a wrapper with the auto-generated class name is applied (a bug resolved by these changes).

__Breaking Change__: (_Deprecation_) All existing block `save` implementations returning raw HTML will continue to work, but a warning will be logged to the browser console recommending the `RawHTML` component instead. We may want to seek to remove this deprecated code in the future.

__Testing instructions:__

Ensure unit tests pass:

```
npm run test-unit
```

Verify that there are no regressions in the behavior of blocks expecting to save raw HTML, notably the Classic, HTML, and Shortcode blocks. These blocks should also not render any wrapper elements, except in the case that a custom class name is assigned.

aduth committed:
Element: Add support for RawHTML component

Explicit support by element static string renderer.

aduth committed:
Element: Render dangerous wrapper if props passed

aduth committed:
Blocks: Update blocks to return RawHTML

aduth committed:
Block API: Drop support for raw HTML save return

aduth commented:
Some additional tasks I'd like to do here:

- Add temporary deprecated backwards compatibility for existing usage
- Update documentation to reflect changes

aduth committed:
Block API: Add hook for deprecated raw HTML support

gziolo commented on blocks/hooks/deprecated.js:
```diff
@@ -0,0 +1,46 @@
+	return <DangerousHTMLWithWarning children={ element } />;
+}
+
+addFilter( 'blocks.getSaveContent.saveElement', 'core/deprecated/save-props', shimDangerousHTML );
```
I like how you handled this backward compatibility issue with the hook ❤️

aduth replied:
>I like how you handled this backward compatibility issue with the hook ❤️

The backwards-compatibility error also had the side benefit of triggering some failing tests which both hinted toward (a) flawed tests and (b) that maybe we want to support _some_ string returns to continue to work unaffected, but not those including raw HTML. This has been improved in the rebased 5cad5d3.

aduth committed:
Serializer: Remove save return from more block test

Unused anyways, misleading to imply it is.

gziolo commented on blocks/api/serializer.js:
```diff
@@ -45,13 +45,10 @@ export function getSaveElement( blockType, attributes ) {
-		}
 	}
 
+	saveElement = applyFilters( 'blocks.getSaveContent.saveElement', saveElement, blockType, attributes );
```
Unrelated to this PR. In the future, we should also make sure that functional components are treated as React components, too:

```js
saveElement = save( { attributes } );
```

gziolo replied:
We are about to introduce new hook `blocks.getSaveContent.saveElement`. Is it meant to exist only for the period of deprecation of `saveElement` which is able to return a string?

aduth replied:
>Unrelated to this PR. In the future, we should also make sure that functional components are treated as React components, too:

Related: https://github.com/WordPress/gutenberg/pull/3745#discussion_r165008918

It is potentially related to this pull request, as we've simplified all paths which necessitated treating the function component any different than the component class. I think at one point in putting this pull request together I'd tried to convert to `createElement`, but ran into challenges. I'll try to revisit and see if it can be done easily.

gziolo replied:
Yes, I tried the similar in the past but ended with splitting into two methods `getSaveContent` and `getSaveElement` only :(

aduth replied:
>We are about to introduce new hook `blocks.getSaveContent.saveElement`. Is it meant to exist only until the `saveElement` is being able to return a string? If the answer is yes, then we might want to move all the deprecation logic here instead of using a hook.

I don't have a problem with it sticking around. Related to #3800, it can serve as useful in cases like this one where we want to override the entire behavior of save, not just inject some additional props.

Of course, this opens questions of when and why we add filters. Also feel that we should be proactive about documenting them (expecting that, as with PHP, we'd want [hook documentation pages](https://developer.wordpress.org/reference/hooks/the_content/).

gziolo commented on blocks/hooks/deprecated.js:
```diff
@@ -0,0 +1,58 @@
+	return element;
+}
+
+addFilter( 'blocks.getSaveContent.saveElement', 'core/deprecated/save-props', shimDangerousHTML );
```
`'core/deprecated/save-props'` -> `'core/deprecated/save-element'`?

aduth replied:
>`'core/deprecated/save-props'` -> `'core/deprecated/save-element'`?

Aside: I still don't like that we name filters 😄  The number of times I've edited this line just to change the name serves as one reason.

gziolo replied:
Yes, this namespace thing (2nd param) seems obsolete at the moment.

gziolo commented on blocks/hooks/deprecated.js:
```diff
@@ -0,0 +1,58 @@
+	// Still support string return from save, but in the same way any component
+	// could render a string, it should be escaped. Therefore, only shim usage
+	// which had included some HTML expected to be unescaped.
+	if ( typeof element === 'string' && includes( element, '<' ) ) {
```
`includes( element, '<' )` - this behavior might lead to some errors. I will comment next to the code which might cause issues.

aduth replied:
>`includes( element, '<' )` - this behavior might lead to some errors. I will comment next to the code which might cause issues.

Which comment expanded on the issue here?

gziolo replied:
https://github.com/WordPress/gutenberg/pull/4786#discussion_r165915137

aduth replied:
Talked with @gziolo about this some more. Basically: We want to support string returns from `save`, but in the same way that string is a valid React element, it should be unescaped. The shim here specifically targets HTML which previous block implementations may have been returning. The one case where this won't work well is if a block had returned a shortcode string, which will now be escaped. We could similarly add support for shortcode patterns, or alternatively just allow them to break (it's unclear to me how common this would have been).

gziolo commented on blocks/api/serializer.js:
```diff
@@ -45,13 +45,10 @@ export function getSaveElement( blockType, attributes ) {
 		saveElement = save( { attributes } );
-
-		// Special-case function render implementation to allow raw HTML return
-		if ( 'string' === typeof saveElement ) {
```
As noted in here https://github.com/WordPress/gutenberg/pull/4786/files#r165907203, we are slightly changing behavior with this PR. When the `saveElement` is going to be represented as a plain string without any HTML tags, this function won't return the value immediately. It will be wrapped with `Children.map`, but `addExtraContainerProps` won't apply filters because it will return immediately after check if an object was passed.  In the case where the `saveElement` would contain HTML tags it is going to be recognized as object by the inner function, therefore all filters will be applied. This might introduce some confusion as the same block can have a different markup depending on the content.

gziolo replied:
Please also note that there is another function in this file:
```js
export function getSaveContent( blockType, attributes ) {
```
which performs check against the return value from `getSaveElement`. It won't return `string` anymore after this refactoring is applied so we can remove the check in lines 78-81. `Children.map` makes sure that return value of the function is always object.

We might also want to explore in another PR the possibility of turning `getSaveElement` in `SaveBlock` component as suggested in #3800 by @andreiglingeanu.

aduth replied:
>In the case where the `saveElement` would contain HTML tags it is going to be recognized as an object by the inner function, therefore all filters will be applied.

If I understand correctly: This is different, yes, but arguably more consistent. Example of this being how the Freeform block could get away with avoiding `supports.className: false` previously only because it returned a `string` from `save` and so it was previously skipped in the [generated class name hook](https://github.com/WordPress/gutenberg/blob/master/blocks/hooks/generated-class-name.js). While in the case of freeform we don't want the generated class name, it seems to me wise to reduce the number of variations in how save results are handled.

gziolo reviewed (changes requested):
This looks good and is almost ready to merge. I have one concern where we need to clarify what is the expected behavior because there might be an unexpected change in the output HTML depending on the content of `RichText` component. I also left some comment where I would like to discuss further refactorings.

aduth committed:
Block API: Document WPBlockType type definition

aduth committed:
Serializer: Simplify serialize support to function

aduth commented:
I've pushed a rebased branch which should address most all of the feedback. I've pushed for more refactoring of the `getSaveElement` function, specifically eliminating explicit support for `Component` class, which was flawed anyways (see fixed "Component save" serializer test case).

I have some stashed changes which pushed even to deprecate `blocks.getSaveContent.extraProps` altogether, instead recommending the newly-introduced `blocks.getSaveElement` which can accomplish the same via the filter handler's own `cloneElement`. However, since this requires the plugin author to accommodate all element shapes (including `null`), I was worried about this being error-prone with implementations trying to call e.g. `element.props.foo` (where `element` is `null`). For now, I've kept it around.

cc @youknowriad I know you'd mentioned on at least one occasion removing support for `save` components.

youknowriad commented on blocks/api/registration.js:
```diff
@@ -16,9 +16,34 @@ import { applyFilters } from '@wordpress/hooks';
+ * Defined behavior of a block type.
  *
- * @type {Object}
+ * @typedef {WPBlockType}
```
We should just use flow or typescript :trollface:

gziolo replied:
Yes, that would even more useful :)

youknowriad commented on blocks/api/serializer.js:
```diff
@@ -37,33 +38,46 @@ export function getBlockDefaultClassname( blockName ) {
 
+	// Component classes are unsupported for save since serialization must
+	// occur synchronously. For improved interoperability with higher-order
+	// components which often return component class, emulate basic support.
```
Good idea 👍

youknowriad commented on blocks/api/test/serializer.js:
```diff
@@ -129,7 +117,7 @@ describe( 'block serializer', () => {
 					{ fruit: 'Bananas' }
 				);
 
-				expect( saved ).toBe( '<div>Bananas</div>' );
```
Do you know why the className was missing here?

aduth replied:
>Do you know why the className was missing here?

Yep, alluded to in https://github.com/WordPress/gutenberg/pull/4786#issuecomment-363133291 with the "fixed test case" mention, but the issue was specifically in how `createElement` is treated differently from calling `save` as a function directly. It's the same reason why we couldn't just refactor to `createElement`:

![image](https://user-images.githubusercontent.com/1779930/35820112-b94e2ab2-0a72-11e8-9146-e26a8970a4d9.png)

...which is that, when cloning the element produced by `createElement`, it's the same as modifying props of what is passed into `save`, _not_ what is returned by `save`. The filtering expects to be applied on whatever is produced by `save`, e.g. the `div` here.

We don't really have a way to hook into the render behavior without some awful machinery to `render` into a temporary DOM, so instead we drop support for `save` as a component and just treat it as a function (i.e. it's still not a "stateless function component", just a function returning an element(s)).

youknowriad commented on blocks/library/freeform/test/__snapshots__/index.js.snap:
```diff
@@ -8,7 +8,7 @@ Array [
     style="display:none"
   />,
   <div
-    class="wp-block-freeform blocks-rich-text__tinymce"
```
Wondering why this change, you didn't touch the `edit` function right?

aduth replied:
>Wondering why this change, you didn't touch the `edit` function right?

Because I added the `supports.className: false`, it's no longer adding the generated class in `BlockEdit`:

https://github.com/WordPress/gutenberg/blob/0cbe106b28270694515930303a8162c5d846bdd0/blocks/block-edit/index.js#L28-L32

...though this makes me think that this is a bit hard to follow, and that we might want to consolidate class name handling into `blocks/hooks/generated-class-name.js`, particularly since `BlockEdit` is already filterable anyways.

youknowriad replied:
Oh right I confused it with `customClassName`

gziolo commented on blocks/api/serializer.js:
```diff
@@ -37,33 +38,46 @@ export function getBlockDefaultClassname( blockName ) {
+		 * @param {Object}      attributes Block attributes.
+		 */
+		const props = applyFilters(
+			'blocks.getSaveContent.extraProps',
```
Should we rename to `blocks.getSaveElement.extraProps`?

aduth replied:
>Should we rename to `blocks.getSaveElement.extraProps` to match with another filter? Can be another PR.

I think we should, yes, but it is a breaking change, one which we can deprecate with a hook, but probably best left for a separate pull request.

gziolo commented on element/index.js:
```diff
@@ -80,9 +86,16 @@ export { createPortal };
+	let rendered = renderToStaticMarkup( element );
+
+	// Drop raw HTML wrappers (support dangerous inner HTML without wrapper)
+	rendered = rendered.replace( /<\/?wp-dangerous-html>/g, '' );
```
Nit: this function can return here instead of updating variable's value.

aduth replied:
>Nit: this function can return here instead of updating variable's value.

I'm going to leave this as-is, as it distinguishes / sets a pattern for transformations which would be applied to the default rendered output.

mtias commented on blocks/library/html/index.js:
```diff
@@ -70,6 +71,6 @@ export const settings = {
 
 	save( { attributes } ) {
-		return attributes.content;
+		return <DangerousHTML>{ attributes.content }</DangerousHTML>;
```
I think this component should be called `RawHTML`, it doesn't make a lot of sense to call it dangerous in the context of an editor that deals with raw html in valid and regular cases.

gziolo replied:
Works for me.

aduth replied:
Yeah, one potential downside is it's basically a substitute for `dangerouslySetInnerHTML` (sans wrapper), not just for serialization, so it could turn into a bit of a foot-gun.

gziolo reviewed (approved):
Code looks good, there are still some minor comments which might need some code updates, but everything looks good. I did a quite long testing session and I didn't discover any regressions when using updated blocks: Classic, Custom HTML, Shortcodes. 👍

aduth commented:
Renamed from `DangerousHTML` to `RawHTML`.

Also added deprecated compatibility support for shortcodes via a regular expression which is prone to false positives (better to capture and warn than break any plugin which had expected to return a shortcode text from its `save`).
